urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/thread/ControlledRunnable.java

```diff
@@ -0,0 +1,753 @@
+	}
+
+	/**
+	 * Removes an existing listener from this class
+	 * 
+	 * @param l
+	 */
+	public void removeControlledRunnableListener(ControlledRunnableListener l) {
+		listeners.remove(l);
+	}
+
+	/**
+	 * Fires the listener with the state of the event
+	 * 
+	 * @param e
+	 */
+	private void fireControlledRunnableListener(ControlledRunnableEvent e) {
+		for (ControlledRunnableListener l : listeners) {
+			l.runnableStateChanged(e);
+		}
+	}
+
+	// ========================================================================
+
+
+	/**
+	 * Returns the flag which indicates if this runnable is currently running or not
+	 * 
+	 * @return
+	 */
+	public boolean isRunning() {
+		return running;
+	}
+
+	// ========================================================================
+
+	/**
+	 * 
+	 * 
+	 * @param timeout
+	 */
+	public void pause(long timeout) {
+		pause(timeout, false);
+	}
+
+	/**
+	 * 
+	 * 
+	 * @param timeout
+	 * @param wait
+	 */
+	public void pause(long timeout, boolean wait) {
+		this.timeout = timeout;
+		pause(true, wait);
+	}
+
+	/**
+	 * Requests the runnable to pause (pause=<code>true</code>) or to exit from
+	 * its paused state (pause=<code>false</code>).
+	 * 
+	 * @param pause
+	 */
+	public void pause(boolean pause) {
+		if (!pausingImplemented) {
+			throw new ControlledRunnableError("Pausing this runnable is not implemented.");
+		}
+
+		this.timeout = 0;
+
+		pauseInternal(pause, false);
+	}
+
+	/**
+	 * Requests the runnable to pause (pause=<code>true</code>) or to exit from
+	 * its paused state (pause=<code>false</code>). With flag to wait for runnable
+	 * to reach the paused state.
+	 * 
+	 * @param pause
+	 * @param wait
+	 */
+	public void pause(boolean pause, boolean wait) {
+		if (!pausingImplemented) {
+			throw new ControlledRunnableError("Pausing this runnable is not implemented.");
+		}
+
+		this.timeout = 0;
+
+		pauseInternal(pause, wait);
+	}
+
+	/**
+	 * Requests the runnable to pause (pause=<code>true</code>) or to exit from
+	 * its paused state (pause=<code>false</code>).
+	 * 
+	 * @param pause
+	 * @param wait
+	 */
+	private void pauseInternal(boolean pause, boolean wait) {
+		if (stopped) {
+			throw new ControlledRunnableError("The pause state of this runnable can not be changed, because it has already been stopped.");
+		}
+
+		//Request pause
+		pauseRequested = pause;
+
+		if (pauseRequested) {
+			fireControlledRunnableListener(
+					new ControlledRunnableEvent(this, ControlledRunnableEvent.STATETYPE.PAUSED));
+		}
+
+		if (!pause && paused) {
+			//No pause any more (if currently paused)
+			synchronized (PAUSESYNC) {
+				PAUSESYNC.notify();
+			}
+		}
+
+		if (wait) {
+			if (pause && !paused || !pause && paused) {
+				//Only synchronize if really necessary
+				synchronized (WAITFORSYNC) {
+					try {
+						//Wait for entering or exiting pause state
+						while (pause && !paused || !pause && paused) {
+							WAITFORSYNC.wait();
+						}
+					} catch (InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)